v4/newrelic/transaction.go

```diff
@@ -257,9 +304,13 @@ func (txn *Transaction) BrowserTimingHeader() *BrowserTimingHeader {
 // Note that any segments that end after the transaction ends will not
 // be reported.
 func (txn *Transaction) NewGoroutine() *Transaction {
+	txn.thread.Lock()
+	defer txn.thread.Unlock()
+
 	newTxn := *txn
 	newTxn.thread = &thread{
-		currentSpan: txn.thread.currentSpan,
+		currentSpan:  txn.thread.currentSpan,
+		isSingleSpan: txn.thread.isSingleSpan,
```

Comment:
Just thinking aloud. I think there's more we need to consider.

```go
package main

import (
	"log"
	"net/http"
	"time"

	"github.com/newrelic/go-agent/v4/newrelic"
	"go.opentelemetry.io/otel/exporters/stdout"
)

func main() {
	pusher, err := stdout.InstallNewPipeline(nil, nil)
	if err != nil {
		log.Fatal(err)
	}
	defer pusher.Stop()

	app, _ := newrelic.NewApplication(
		newrelic.ConfigFromEnvironment(),
	)
	txn := app.StartTransaction("main")
	defer txn.End()

	segFunc := func(name string, txn *newrelic.Transaction) {
		defer txn.StartSegment(name).End()
	}

	go segFunc("seg1", txn.NewGoroutine())
	time.Sleep(time.Second)

	go func(txn *newrelic.Transaction) {
		hdr := http.Header{}
		hdr.Set("traceparent", "00-aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa-bbbbbbbbbbbbbbbb-01")
		txn.AcceptDistributedTraceHeaders("HTTP", hdr)
		segFunc("seg2", txn)
	}(txn.NewGoroutine())
	time.Sleep(time.Second)

	segFunc("seg3", txn)
}
```

The first segment and the third segment will be reported, but the second one will be orphaned. This is because the new root span that is created when Accept is called is never ended. When `txn.End()` is called, it is ending and finalizing the original root span.

Possible solution: disallow swapping of root spans if Accept is called inside a goroutine (aka `txn.NewGoroutine` was the creator of the `txn.thread`).

Thoughts?

Reply:
Good point. Is it a common practice (or even allowed) to call `txn.End` on the transaction that was returned from `NewGoroutine`? Or do we anticipate that this call always happens on the original transaction?

Reply:
We have always touted our `*Transaction` as being thread safe and thus that you can call `txn.End` anywhere :(

Reply:
> Possible solution: disallow swapping of root spans if Accept is called inside a goroutine (aka txn.NewGoroutine was the creator of the txn.thread).

I think to be safe we have to go a little bit further and disallow root span swapping in general once `NewGoroutine` was called, for the original transaction as well as for the one created via `NewGoroutine`. We have to prevent that transactions have different root segments, otherwise calls to `txn.End` might cause orphaned spans.

Reply:
I pushed a commit implementing that.